M2/Macaulay2/packages/Macaulay2Doc/functions/loadProgram-doc.m2

```diff
@@ -40,27 +45,48 @@ document {
 document {
     Key => {loadProgram,
 	(loadProgram, String, String),
+	(loadProgram, String, List),
+	[loadProgram, Prefix],
 	[loadProgram, RaiseError],
 	[loadProgram, Verbose]},
     Headline => "load external program",
-    Usage => "loadProgram(name, cmd)",
+    Usage => "loadProgram(name, cmd)\nloadProgram(name, cmds)",
     Inputs => {
 	"name" => String => {"the name of the program to load.  ",
 	    "This should match the corresponding key in ",
 	    TO "programPaths", "."},
 	"cmd" => String =>
 	    "a command to run that should return 0 if the program is present.",
+	"cmds" => List => {"a list of commands to run that should all return ",
+	    "0 if the program is present."},
 	RaiseError => Boolean =>
 	    "whether to raise an error if the program is not found.",
 	Verbose => Boolean =>
 	    "whether to inform the user of each path that is checked.",
+	Prefix => List => {
+	    "a list of sequences containing three strings identifying a ",
+	    "prefix that is added to the executable binaries belonging to the ",
+	    "program by some distributions.  These sequences should be of the ",
+	    "form ", TT "(distribution, regex, prefix)", " where: ",
+		UL {
+		    {TT "distribution", " identifies the name of the ",
+			"distribution that uses the prefix, e.g., ",
+			"\"debian\".  It is only used for printing messages ",
+			"when ", TT "Verbose", " is set to ", TO "true", "."},
```

Comment:
"distribution" is vague.  What about defining it to be the initial alphanumeric part of `version#"issue"`?  We could even add an entry `version#"distribution"` based on that definition.

Reply:
The idea is that the package author would set this when they write a package that interfaces with an external program.  They'd need to do the legwork to see what the various prefixes are in various distributions.  It shouldn't be automatically generated.  For example:

```M2
	topcomProgram = loadProgram("topcom","cube 3", Prefix => {
	    ("debian", ".*", "topcom-"),
	    ("fedora", "^(cross|cube|cyclic|hypersimplex|lattice)$", "TOPCOM-"),
	    ("gentoo", "^cube$", "topcom_")});
```

Reply:
But how does the author know whether to write "debian" or "Debian" for the distribution?  That's what's vague.

Reply:
Also vague is whether, under Ubuntu, to write "Ubuntu" or "Debian".

Reply:
It doesn't really matter, as it's just there to add some readability to the verbose output.  `loadProgram` just tries every path/prefix combination by brute force until it finds one that works.  It doesn't actually try to detect what the system actually is.  For example:

```m2
i3 : loadProgram("topcom", "cube 3", Verbose => true, Prefix => {
         ("debian", ".*", "topcom-"),
         ("fedora", "^(cross|cube|cyclic|hypersimplex|lattice)$", "TOPCOM-"),
         ("gentoo", "^cube$", "topcom_")})
checking for topcom in /usr/libexec/Macaulay2/bin/...
         trying default prefix...
        not found
         trying debian prefix...
        not found
         trying fedora prefix...
        not found
         trying gentoo prefix...
        not found
checking for topcom in /home/profzoom/.local/bin/...
         trying default prefix...
        not found
         trying debian prefix...
        not found
         trying fedora prefix...
        not found
         trying gentoo prefix...
        not found
checking for topcom in /usr/local/sbin/...
         trying default prefix...
        not found
         trying debian prefix...
        not found
         trying fedora prefix...
        not found
         trying gentoo prefix...
        not found
checking for topcom in /usr/local/bin/...
         trying default prefix...
        not found
         trying debian prefix...
        not found
         trying fedora prefix...
        not found
         trying gentoo prefix...
        not found
checking for topcom in /usr/sbin/...
         trying default prefix...
        not found
         trying debian prefix...
        not found
         trying fedora prefix...
        not found
         trying gentoo prefix...
        not found
checking for topcom in /usr/bin/...
         trying default prefix...
        not found
         trying debian prefix...
        found
```

I could put a suggestion in the documentation as to what it should be, e.g., the "parent" distribution in lowercase or something, so that packages are consistent.

Reply:
Oh!  It would be better, instead of demanding what is essentially just a comment from the user, to just display the prefix itself that you're trying.  It will give complete information, after all.

Reply:
Ah, gotcha.  I think my thought was that some distributions may end up using the same prefix, but not on the same binaries.  Say, Debian ships `foo-bar` and `foo-baz`, but Fedora ships `foo-bar` and `baz`.  So just the prefix may not be sufficient to disambiguate the different cases.

This may never happen in reality, though.  I'll double-check nauty and 4ti2 to see what the different distributions use as prefixes.  If it's not an issue, then I'll drop that first string.

Reply:
To handle the case you're worried about, when looking for `bar`, look for both `foo-bar` and `bar`.  Then also, when looking for `baz`, look for both `foo-baz` and `baz`.  That makes the regular expression redundant, too.

Reply:
If we remove the regular expression, then `runProgram` would have to worry about trying both things, too.  My goal was for `loadProgram` to deal with everything up front so that `runProgram` would have it easy.

Reply:
It looks like the nauty and 4ti2 prefixes are all more sane than the topcom ones (either everything is prefixed with `nauty-`, `4ti2-`, or `4ti2_`, or nothing is), so it should be fine to drop the distribution string.

Whenever I figure out the failing workflow, I'll push a couple more commits.

Reply:
Even if we do run into that case at some point, I suppose it's not the end of the word if the verbose output looks a little redundant, even though behind the scenes the checks are slightly different:

```m2
checking for foo in /usr/bin/...
         trying prefix ""...
        not found
         trying prefix "foo-"...
        not found
         trying prefix "foo-"...
        not found
```

Reply:
Or even better, I'll just include the regex in the message, too.

Reply:
> If we remove the regular expression, then `runProgram` would have to worry about trying both things, too. My goal was for `loadProgram` to deal with everything up front so that `runProgram` would have it easy.

Not if `runProgram` records where each program was found, so `runProgram` knows where and under what name it was found.

Reply:
> > If we remove the regular expression, then `runProgram` would have to worry about trying both things, too. My goal was for `loadProgram` to deal with everything up front so that `runProgram` would have it easy.
> 
> Not if `loadProgram` records where each program was found, so `runProgram` knows where and under what name it was found.

But `loadProgram` currently doesn't check each program -- only the minimal number of them needed to figure out the path and prefix.  Even for TOPCOM, we only need to call one thing -- the prefix of `cube` gives us enough information to figure out all of the prefixes, provided we know where to get the information, which is what the regexes are doing.

TOPCOM comes with 42 little programs, so I'd rather not check each one for each prefix/path pair.